flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java

```diff
@@ -186,6 +186,8 @@ public void addShipFiles(List<File> shipFiles) {
 		for (File shipFile: shipFiles) {
 			// remove uberjar from ship list (by default everything in the lib/ folder is added to
 			// the list of files to ship, but we handle the uberjar separately.
+			// NOTE: this does not filter the uberjar if added recursively, e.g. by having the "lib"
+			//       folder in the shipFiles
 			if (!(shipFile.getName().startsWith("flink-dist") && shipFile.getName().endsWith("jar"))) {
```

Comment:
this filtering seems redundant now, the non-directory branch in `uploadAndRegisterFiles` should catch it.

We may want to pull the directory expansion into this method instead, and keep uploadAndRegisterFiles as a dumb upload method.

Reply:
oh - you're right.

Regarding the change of `uploadAndRegisterFiles`: I wouldn't upload the shipfiles in `addShipFiles` yet, so if we changed that, we'd have to traverse through the directories twice as we definitely need to do that for the uploads and if we wanted to filter recursively in `addShipFiles`, too, then there as well.